Assets/Scripts/Inventory/Inventory.cs

```diff
@@ -2,65 +2,148 @@
 using System.Collections.Generic;
 using UnityEngine;
 
-// This class represents a slot within the player's inventory that holds an item
-[System.Serializable]
-public class ItemSlot
+namespace FunMath
 {
-    public ItemSlot(Item item, int Count)
+    // This class represents a slot within the player's inventory that holds an item
+    [System.Serializable]
+    public class ItemSlot

```

Comment:
I still don't get the point of why we need ItemSlot. If we create a list (or two lists), isn't the index gonna indicate the slot of an item? Why does a slot contain count? Doesn't it only apply to NumberItem? What if we pick up 10 addition operators or 20 division operators? I thought they were only types. 

Here is my 2 cents of creating an inventory another associated classes:
```
Enum Operator
{
  Addition,
  Subtraction,
  Multiplication,
  Division,
  ...
}

Class NumberItem
{
  int value;
  int count;

  private bool IsEmpty() { ... }
}

Class Inventory
{
  List of Operators
  List of NumberItems

  // Here is the interface of the inventory
  Operator GetCurrentOperator(); // returns the current operator
  Operator SelectPreviousOperator(); // move idx to prev & returns the current operator
  Operator SelectNextOperator(); // return idx to next & returns the current operator
  NumberItem GetNumberItem(int val); // return null if not exist, return the object otherwise
  void ModifyNumberItem(NumberItem item); // add new if not exist, modify count otherwise
}
```

Reply:
> I still don't get the point of why we need ItemSlot. If we create a list (or two lists), isn't the index gonna indicate the slot of an item?

I'm fine with the `SlotItem` as a way to hold the data about the slot.  This is more meaningful when we allow the ability to drop items and manage inventory.  Since you won't be able to rely on the index to match with an enum.  So you need something to tell you, this slot in the list of operators is occupied by the `Addition` operator and it has `X` remaining count.
Think of it like the following JSON array
```json
"inventory": [{"type": "Addition", "count": 5}, {"type": "Multiplication", "count": 1}]
```

Reply:
> `Class Inventory
{
  List of Operators
  List of NumberItems ...`

I'd also push back on the Inventory being considered as the one place to find all the different types of items held by the player (you have it containing separate lists and functions for getting out numbers or operators).  I like the reusability of having an inventory dedicated to one type of item either operator **XOR** modifier (number).  This way we define one class that can be reused like so
```csharp
class Inventory<Item<T>>
{
  private List<Item<T>> Slots;
  Item<T> GetItemAt(int index);
  void ModifyItemCount(Item<T> item); // Item has a type and a count as per my last comment, so we could input one with a positive value to add some and negative to remove some
  event OnCountModified(int index, Item<T> item);  // The event tells you which item is changing and gives the index so the UI doesn't have to re-implement anything to find which display slot to update
}
```
This means the inventory is in charge of one thing: keeping stock of its items.  If we want, we can add the ability to track which item is actively selected.  If so, we don't need the `InventorySelector` class that was introduced in this PR.  But if you want to keep that abstraction, which I think you should, then the `InventorySelector` has the simple job of keeping track of which index is selected and then to decrement the `Inventory` at that index whenever something is used.  

Note: the player will still need two inventory objects, but things that need them will know which one to grab and how to change it's current index etc.

Reply:
@LuffaKing 

> I am still struggling to understand the purpose of ItemSlot. If we create a list (or two lists), doesn't the index inherently indicate the slot of an item? 

Item Slot is an abstraction that allows the UI to answer this simple question: "What if the user hasn't discovered an item?"
It actually does a lot more. I've done multiple inventory systems and I've been burnt by removing this small abstraction before. The biggest win is that the interaction with UI is more explicit.

>Why does a slot contain count 
The original idea had operators that minus count, so this was a remnant of it. It works out great because we can reuse the same code and type in a data abstracted format.

>What if we collect 10 addition operators or 20 division operators? I was under the impression that these were merely types.
It really wouldn't matter how many we collect once that infinity checkbox is clicked. The number will never move from there.

I hope this answers your questions.

Reply:
Right, I was wrong about the count for the operators. Yes, we need the count for that. 

@logar16 
> the player will still need two inventory objects, but things that need them will know which one to grab and how to change it's current index etc.

I am not sure if we need two inventory instances. If we were to introduce an additional list, it would imply the need for another inventory instance, it sounds a bit overhead to me. Having an index for each list is cheap. I don't see significant drawbacks to having one inventory that has all the information about collected items. 

>But as I write this, I'm thinking maybe the ItemSlot and Item can be combined.
I agree with this. 

@ShumWengSang 
> Item Slot is an abstraction that allows the UI to answer this simple question: "What if the user hasn't discovered an item?"

I don't see a huge benefit to having ItemSlot for now. Regarding the UI, assuming the max number of operators is 5, If the user hasn't collected any items, we can display 5 empty slots (null) on the UI. I would like to know more about the disadvantages of not using ItemSlot. 

Overall, I don't disagree with the ideas you suggested. As long as I can call the method(s) to get the information I need for the attack mechanism.